feat(freetype): add colored glyphs support

Moschn opened the pull request:
### Freetype: Support Colored Bitmap Glyphs

Fonts often come with bitmap glyphs, e.g. for emojis, and Freetype already supports these glyphs. This PR enables LVGL to receive colored bitmaps from Freetype and draw them. 

Since this is my first contribution to LVGL, I am probably making many mistakes. I am open for feedback. 

Notes:

- The drawing of ARGB888 glyphs is not yet define guarded. This should probably be guarded, but I was not sure which one to choose. Also, the code is currently copy pasted from `LV_FONT_GLYPH_FORMAT_IMAGE`. Ideally, we would not duplicate code, but I did not figure out a way to set the `glyph_draw_dsc->format` correctly in `lv_freetype_image.c`. Let me know if you know of a way to correctly tag a glyph in `lv_freetype_image.c` with `LV_FONT_GLYPH_FORMAT_IMAGE`.
- Documentation could also be updated to help users use this new feature. I added an example already, but was not sure if I should also add documentation (and where).
- Ideally, LVGL would also support drawing SVG colored glyphs as those can easily be scaled. I believe many emoji fonts such as twemoji and newer versions of NotoColoredEmoji use scalable glyphs in this way. Adding such a feature would require quite a bit of knowledge in LVGL and Freetype though.

W-Mai commented on src/libs/freetype/lv_freetype_glyph.c:
```diff
@@ -136,8 +136,19 @@ static bool freetype_glyph_create_cb(lv_freetype_glyph_cache_data_t * data, void
 
-    FT_Set_Pixel_Sizes(face, 0, dsc->size);
-    error = FT_Load_Glyph(face, glyph_index,  FT_LOAD_COMPUTE_METRICS | FT_LOAD_NO_BITMAP | FT_LOAD_NO_AUTOHINT);
+    if(FT_IS_SCALABLE(face)) {
+        error = FT_Set_Pixel_Sizes(face, 0, dsc->size);
+    }
+    else {
+        LV_LOG_WARN("font is not scalable, selecting available size\n");
+        error = FT_Select_Size(face, 0);
+    }
+    if(error) {
+        FT_ERROR_MSG("FT_Set_Pixel_Sizes", error);
+        lv_mutex_unlock(&dsc->cache_node->face_lock);
+        return false;
+    }
+    error = FT_Load_Glyph(face, glyph_index,  FT_LOAD_COMPUTE_METRICS | FT_LOAD_NO_AUTOHINT);
```
You can change the g_dsc->format = LV_FONT_GLYPH_FORMAT_IMAGE

Moschn replied:
Thanks for the hint. So the format can be selected when loading the glyph in `lv_freetype_glyph`, but at that point, I am not sure if we know yet if it is a colored bitmap. 

TBH, I did not yet fully understand the flow. First the glyph is loaded:
https://github.com/lvgl/lvgl/blob/d7b82c79668d7b31d52f6535360eff15396eade5/src/libs/freetype/lv_freetype_glyph.c#L140
but then the glyph is loaded again but with different flags?
https://github.com/lvgl/lvgl/blob/d7b82c79668d7b31d52f6535360eff15396eade5/src/libs/freetype/lv_freetype_image.c#L131
Adding prints/or using a GDB shows that both paths are being hit when rendering a freetype font.

W-Mai commented:
<img width="1313" alt="image" src="https://github.com/user-attachments/assets/4ed42fd0-a275-4042-ad42-e71dbf7dede1">

Sorry but I can't reproduce your code. I got a error message like this:

<img width="856" alt="image" src="https://github.com/user-attachments/assets/858d4a67-b39c-4cd4-9843-65092f02b335">

<img width="1023" alt="image" src="https://github.com/user-attachments/assets/28e0afaa-b308-4629-abc8-8163e1ff6feb">

So, what configuration information did I miss?

W-Mai commented:
Well, when I open this option, it works well

<img width="487" alt="image" src="https://github.com/user-attachments/assets/9cb97352-9fbf-4bb2-930a-f027358d06c7">

<img width="485" alt="image" src="https://github.com/user-attachments/assets/314f216c-ddee-44f3-a8f0-6cc411014963">

kisvegabor commented on examples/libs/freetype/lv_example_freetype_2.c:
```diff
@@ -0,0 +1,62 @@
+    lv_obj_t * label = lv_label_create(lv_screen_active());
+    lv_obj_add_style(label, &style, 0);
+    lv_label_set_text(label, "Hello world\nI'm a font created with FreeType 😀");
+    lv_obj_center(label);
```
Please add something similar [as a test](https://github.com/lvgl/lvgl/blob/master/tests/src/test_cases/libs/test_freetype.c) as well.

kisvegabor reviewed (commented):
It's a great update! :rocket: 

@W-Mai can comment more on the technical details, I can point out some high-level things:
- Please add a test
- Please add a paragraph about it [in the docs](https://github.com/lvgl/lvgl/blob/master/docs/libs/freetype.rst)

Moschn commented:
> Well, when I open this option, it works well
> <img alt="image" width="487" src="https://private-user-images.githubusercontent.com/24841247/358120384-9cb97352-9fbf-4bb2-930a-f027358d06c7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM3MDgyNTcsIm5iZiI6MTcyMzcwNzk1NywicGF0aCI6Ii8yNDg0MTI0Ny8zNTgxMjAzODQtOWNiOTczNTItOWZiZi00YmIyLTkzMGEtZjAyNzM1OGQwNmM3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE1VDA3NDU1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMxMDBlNTI1MjgzNjRiY2UxMzk4M2VjNzYzOGZlNGIyMWI2OTI4N2QxYzI5YzhiMjg4MDY1MWM1NzEzZDBiMWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.HMXzch255zVU03T9mbfEPvLGwXeVALgYbJTuXGUzFlY"> <img alt="image" width="485" src="https://private-user-images.githubusercontent.com/24841247/358120418-314f216c-ddee-44f3-a8f0-6cc411014963.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM3MDgyNTcsIm5iZiI6MTcyMzcwNzk1NywicGF0aCI6Ii8yNDg0MTI0Ny8zNTgxMjA0MTgtMzE0ZjIxNmMtZGRlZS00NGYzLWE4ZjAtNmNjNDExMDE0OTYzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE1VDA3NDU1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJkMGQ1ZmNlYWJhM2QzMTJiOTFlNjliYzlhYzJiYTA2NzU1MDFjYjA5ZThmYjU3ZGU3ODYwZDVjNmQ3N2I4MGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.RxzyOfw-1wEzoiS1UdPvBtZy2qpsMf0gh1wikxfub4g">

Oh very interesting. I compiled freetype2 from upstream but I did not change any of its configuration. I will add a log message that points to this flag such that users can find it more easily.

@kisvegabor Thanks for the pointers. I will add a test and documentation.

Moschn commented:
I added some documentation, and added test coverage.

I now just tagged all bitmap glyphs from FreeType to be `LV_FONT_GLYPH_FORMAT_IMAGE`. This seems to work.

Drawing a letter as an image is guarded by `LV_USE_IMGFONT`. I guess the feature in this PR should work even when `LV_USE_IMGFONT=0`. Therefore, I removed the define guard in `src/draw/sw/lv_draw_sw_letter.c`. I could also add `|| LV_FREETYPE` to the define guard. What do you prefer?

W-Mai commented on tests/src/test_files/fonts/noto/NotoColorEmoji-32.subset.ttf:
One copy is ok, other path can be loaded from a relative path

Moschn commented:
I removed the duplicated font files. Any further comments?

kisvegabor commented on tests/src/test_cases/libs/test_freetype.c:
```diff
@@ -456,6 +466,16 @@ void test_freetype_bitmap_rendering_test(void)
     lv_label_set_text(label2, UNIVERSAL_DECLARATION_OF_HUMAN_RIGHTS_JP);
     lv_obj_align_to(label2, label1, LV_ALIGN_OUT_BOTTOM_MID, 0, 10);
 
+    // test emoji rendering
+    // emoji font does not contain normal characters, use fallback to render them
```
```suggestion
    /* test emoji rendering
     * emoji font does not contain normal characters, use fallback to render them */
```

kisvegabor commented on examples/libs/freetype/lv_example_freetype_2.c:
```diff
@@ -0,0 +1,62 @@
+                                               24,
+                                               LV_FREETYPE_FONT_STYLE_NORMAL);
+
+    // this font is created from a downscaled NotoColorEmoji to 34x32px
+    // Subset containing only a single emoji was created using fonttools:
+    // Command: fonttools subset NotoColorEmoji.ttf --text=😀
```
```suggestion
    /* this font is created from a downscaled NotoColorEmoji to 34x32px
     * Subset containing only a single emoji was created using fonttools:
     * Command: fonttools subset NotoColorEmoji.ttf --text=😀 */
```

kisvegabor reviewed (approved):
It looks good to me, thank you! 

I've just changed `//` to  `/* */`

W-Mai commented on src/libs/freetype/lv_freetype.c:
```diff
@@ -161,7 +161,18 @@ lv_font_t * lv_freetype_font_create(const char * pathname, lv_freetype_font_rend
+    if(FT_IS_SCALABLE(face)) {
+        error = FT_Set_Pixel_Sizes(face, 0, size);
+    }
+    else {
+        LV_LOG_WARN("font is not scalable, selecting available size\n");
+        error = FT_Select_Size(face, 0);
+    }
+    if(error) {
+        FT_ERROR_MSG("FT_Set_Pixel_Sizes", error);
+        return false;
+    }
```
```suggestion
    else {
        LV_LOG_WARN("font is not scalable, selecting available size");
        error = FT_Select_Size(face, 0);
    }
    if(error) {
        FT_ERROR_MSG("FT_Set_Pixel_Sizes", error);
        return NULL;
    }
```

W-Mai replied:
```suggestion
    else {
        LV_LOG_WARN("font is not scalable, selecting available size");
        error = FT_Select_Size(face, 0);
    }
    if(error) {
        FT_ERROR_MSG("FT_Set_Pixel_Sizes", error);
        return NULL;
    }
```

W-Mai commented:
```cpp
typedef enum {
    LV_FREETYPE_FONT_RENDER_MODE_BITMAP = 0,
    LV_FREETYPE_FONT_RENDER_MODE_OUTLINE = 1,
} lv_freetype_font_render_mode_t;
```

You can add this enum as you wish. Like `LV_FREETYPE_FONT_RENDER_MODE_IMAGE`.

Do like this will keep some branch more clearer I think.

Moschn commented:
Thanks for the feedback. I added the changes that you suggested. 

I did not see how adding an extra entry into `lv_freetype_font_render_mode_t` would help simplify some branches/code. But if you have a particular branch, I would be happy to refactor that with a new entry in `lv_freetype_font_render_mode_t`.

W-Mai commented:
> Thanks for the feedback. I added the changes that you suggested.
> 
> I did not see how adding an extra entry into `lv_freetype_font_render_mode_t` would help simplify some branches/code. But if you have a particular branch, I would be happy to refactor that with a new entry in `lv_freetype_font_render_mode_t`.

Now is perfect 😁

liamHowatt reviewed (approved):
Nice! Looks great to me!
